Add a YAML MetaFactory

obitech opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Add an external MetaFactory that allows for serializing multiple YAML (or JSON) files in kubeadm. 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #80941

**Special notes for your reviewer**:
I'm carrying over @luxas work from #78173 , please let me know if this is missing some functionality / tests !

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/assign @stealthybox 
/assign @sttts
/assign @lavalamp

luxas committed:
Add a YAML MetaFactory

- Use `schema.TypeMeta` instead of custom `struct` for VK
- More strict check on GVK after `Interpret` in `SplitYAMLDocuments`
- Adjust `Interpret` comment to include JSON

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go:
```diff
@@ -0,0 +1,54 @@
+		Kind string `json:"kind,omitempty"`
+	}{}
+	if err := yaml.Unmarshal(data, &findKind); err != nil {
+		return nil, fmt.Errorf("couldn't get version/kind; json parse error: %v", err)
```
`return nil, fmt.Errorf("could not interpret GroupVersionKind; unmarshal error: %v", err)`

obitech replied:
Fixed.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go:
```diff
@@ -0,0 +1,54 @@
+// Interpret will return the APIVersion and Kind of the JSON wire-format
+// encoding of an object, or an error.
+func (SimpleMetaFactory) Interpret(data []byte) (*schema.GroupVersionKind, error) {
+	findKind := struct {
```
this should be a TypeMeta instance instead:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/types.go#L36-L41

`typeMeta := runtime.TypeMeta{}`

obitech replied:
Fixed.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go:
```diff
@@ -0,0 +1,128 @@
+	}
+}
+
+func TestInterpret(t *testing.T) {
```
the existing pattern for unit tests is slightly different:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version_test.go#L138-L161

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/scheme_test.go#L613

- define the test cases as a slice of structs as a inline element of the Test* function.
- all test cases should have names/descriptions - e.g. `name: successfully interpret GVK`.
- iterate over all test cases and call Interpret() tests on them.

obitech replied:
Ah thanks for pointing me towards that! I adjusted the test cases and I like them much better now.

neolit123 reviewed (commented):
thanks for the PR @obitech 
i've added some comments.

neolit123 commented:
/priority important-longterm
/ok-to-test

neolit123 commented:
/test pull-kubernetes-e2e-kind

obitech commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta.go:
```diff
@@ -0,0 +1,54 @@
+// DefaultMetaFactory is a default factory for versioning objects in JSON. The object
+// in memory and in the default JSON serialization will use the "kind" and "apiVersion"
+// fields.
+var DefaultMetaFactory = SimpleMetaFactory{}
```
I adjusted the comment.

obitech commented on cmd/kubeadm/app/util/marshal.go:
```diff
@@ -92,31 +91,19 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
-			continue
+		gvk, err := yamlserializer.DefaultMetaFactory.Interpret(b)
+		if err != nil {
+			return nil, errors.Wrapf(err, "invalid configuration: kind and apiVersion is mandatory information that needs to be specified in all YAML documents")
```
So you meant to leave the `schema.ParseGroupVersion` in `Interpret` but add an additional check in `SplitYAMLDocuments` ? I've done it like that now.

obitech commented on cmd/kubeadm/app/util/marshal.go:
```diff
@@ -80,7 +80,6 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
 	buf := bytes.NewBuffer(yamlBytes)
 	reader := utilyaml.NewYAMLReader(bufio.NewReader(buf))
```
Ok maybe I can do this in a follow-up PR?

neolit123 replied:
you can bring this in a WG component-standard meeting for discussion if you like.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go:
```diff
@@ -0,0 +1,111 @@
+	}
+
+	for _, test := range testCases {
+		actual, err := DefaultMetaFactory.Interpret([]byte(test.input))
```
please run these as sub-tests.
t.Run(...) as the examples i gave before.

obitech replied:
Ah I missed that. Fixed now.

neolit123 commented on cmd/kubeadm/app/util/marshal.go:
```diff
@@ -92,31 +91,23 @@ func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, e
-		if len(typeMetaInfo.APIVersion) == 0 || len(typeMetaInfo.Kind) == 0 {
-			errs = append(errs, errors.New("invalid configuration: kind and apiVersion is mandatory information that needs to be specified in all YAML documents"))
-			continue
+		if len(gvk.Version) == 0 || len(gvk.Kind) == 0 {
```
please check gvk.Group too.

obitech replied:
Added.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go:
```diff
@@ -0,0 +1,111 @@
+		switch {
+		case test.errFn != nil:
+			if !test.errFn(err) {
+				t.Errorf("unexpected error returned with test '%s', got: %v", test.name, err)
```
'%s' -> %q

obitech replied:
I included the name of the test as the actual test name in `t.Run()`, so I removed the print of the name here.

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go:
```diff
@@ -0,0 +1,111 @@
+				t.Errorf("unexpected error returned with test '%s', got: %v", test.name, err)
+			}
+		case err != nil:
+			t.Errorf("unexpected error with test '%s': %v", test.name, err)
```
[1]

neolit123 commented on staging/src/k8s.io/apimachinery/pkg/runtime/serializer/yaml/meta_test.go:
```diff
@@ -0,0 +1,111 @@
+		case err != nil:
+			t.Errorf("unexpected error with test '%s': %v", test.name, err)
+		case !reflect.DeepEqual(test.expected, actual):
+			t.Errorf("outcome mismatch with test '%s' -- expected %v, actual: %v",
```
[1]

neolit123 commented:
thanks for the update, added some more comments.

rosti reviewed (commented):
Thanks @obitech for picking this up!
Is there a doc or an issue, that explains in greater detail what we are trying to achieve here? This looks like the first step of a bigger effort and I wonder how this is supposed to continue.
Perhaps @luxas or @stealthybox have more info here?

neolit123 commented:
/assign @stealthybox

neolit123 commented:
/assign @rosti

neolit123 commented:
@obitech 
forgot to mention, please squash the commits into one and force-push.

obitech commented:
Squashed!

/test pull-kubernetes-kubemark-e2e-gce-big

rosti commented:
@obitech , @stealthybox what's the plan for continuing this work in the future? Are we going to extract `SplitYAMLDocuments`? What's the end result, that we try to achieve here?

obitech commented:
@rosti I've put this up as an agenda item for tomorrow's component-standard meeting where I would like to discuss the road map and get more information about the bigger picture. I'll update here afterwards.

stealthybox commented:
[Tues Aug13 WG meeting notes](https://docs.google.com/document/d/18TsodX0fqQgViQ7HHUTAhiAwkf6bNhPXH4vNVTI7GwI/edit#heading=h.n3iz1rg9hvxt)

Would like to review other multi-doc support in existing code other than kubeadm
- Kubectl multi-doc
- Kustomize stuff
- Anything using splitYamlDocuments() in yaml decoder by proxy of an Exported function?
https://github.com/kubernetes/apimachinery/blame/master/pkg/util/yaml/decoder.go#L142

obitech commented:
/remove-lifecycle stale

luxas commented:
ping @sttts @liggitt @lavalamp @mtaufen  for review and approval.
@obitech needs a small autogenerated BUILD file rebase.
Otherwise this is ready to merge

liggitt commented:
/approve

serializer/yaml change lgtm

obitech commented:
Updated, rebased and squashed.

@luxas thanks for your comment on this! I believe there were still some questions in the discussion by @rosti what the future plans are for this effort. Since this started some time ago and I just carried it over, were there any ideas on how to build upon this patch to serialize multiple YAML files? Can we use this to be used inside a `CodecFactory`, for example?

luxas commented:
> @rosti what the future plans are for this effort. Since this started some time ago and I just carried it over, were there any ideas on how to build upon this patch to serialize multiple YAML files? Can we use this to be used inside a CodecFactory, for example?

We still need it, but it's nothing that will be used by an "end-user" of API machinery. Instead, it'll be used internally for higher-level Serializer/CodecFactory-style abstractions that are able to e.g. read a YAML file and directly assess what type it is, **without** marshalling the full type into memory. You can take a look at my ConfigSerializer prototypes in earlier PRs showcasing this: #72883

stealthybox commented:
Just talked about this in kubeadm triage.
This patch is looking good.

We recall there was some discussion of how generally useful this abstraction is.

Alex did some work to identify how multi-doc is handled in other places such as kubectl, and we found other points of interest in the multiple YAML Unmarshal code-paths that exist across the project.

@obitech, can we dig up that work before we decide to merge this?
I personally think this refactor is still a useful step toward working with multi-doc and ComponentConfig.

I'll schedule this for discussion on the next wg-component-standard call.

obitech commented:
> @obitech, can we dig up that work before we decide to merge this?

I've documented some steps here: https://github.com/obitech/k8s-yaml-test

dims commented:
/lgtm
